### PR TITLE
feat: token refresh implemented

### PR DESCRIPTION
**Pull request by pb82**
Implements the oidc token refresh for Observatorium authentication

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -34,23 +63,70 @@ func main() {
...
+		err = remotewrite.ValidateRequest(remoteWriteRequest)
+		if err != nil {
+			log.Printf("error validating the remote write request: %v", err)
 			w.WriteHeader(http.StatusForbidden)
-		} else {
-			proxy.ServeHTTP(w, r)
 		}
+
+		// copy the remote write request back onto the http request
+		err = populateRequestBody(remoteWriteRequest, r)
```
Comment (machi1990) [inline main.go:112]: This is needed because the body has been read already?

File: main.go [hunk 2]
```diff
@@ -34,23 +63,70 @@ func main() {
...
+		if err != nil {
+			log.Printf("error copying remote write request: %v", err)
+			w.WriteHeader(http.StatusInternalServerError)
```
Comment (machi1990) [inline main.go:115]:
```suggestion
			w.WriteHeader(http.StatusInternalServerError)
			return;
```

File: main.go [hunk 3]
```diff
@@ -34,23 +63,70 @@ func main() {
...
+		}
+
+		log.Printf("remote write request received")
+
+		// validate the cluster ids contained in the remote write request
+		err = remotewrite.ValidateRequest(remoteWriteRequest)
+		if err != nil {
+			log.Printf("error validating the remote write request: %v", err)
 			w.WriteHeader(http.StatusForbidden)
```
Comment (machi1990) [inline main.go:107]:
```suggestion
			w.WriteHeader(http.StatusForbidden)
			return
```

File: main.go [hunk 4]
```diff
@@ -34,23 +63,70 @@ func main() {
...
+		proxy.Transport = http.DefaultTransport
+	}
 
 	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
-		err := remotewrite.ValidateRequest(r)
+		// extract the remote write request from the http request
+		remoteWriteRequest, err := remotewrite.DecodeWriteRequest(r)
 		if err != nil {
+			log.Printf("error decoding remote write request: %v", err)
+			w.WriteHeader(http.StatusInternalServerError)
```
Comment (machi1990) [inline main.go:97]:
```suggestion
			w.WriteHeader(http.StatusInternalServerError)
			return
```

File: main.go [hunk 5]
```diff
@@ -34,23 +63,70 @@ func main() {
...
 		}
+
+		// copy the remote write request back onto the http request
+		err = populateRequestBody(remoteWriteRequest, r)
+		if err != nil {
+			log.Printf("error copying remote write request: %v", err)
+			w.WriteHeader(http.StatusInternalServerError)
+		}
+
+		// TODO validate data plane auth token
```
Comment (machi1990) [inline main.go:119]: FYI: This can be doable already. KFM exposes an endpoint for this verification.

File: main.go (hunk 4 above)
Comment (machi1990) [reply main.go:97]: There is no need to execute other part of the code if the error occurs

File: main.go (hunk 3 above)
Comment (machi1990) [reply main.go:107]: No need to remove write if there is an error

File: main.go (hunk 2 above)
Comment (machi1990) [reply main.go:115]: no need to proxy the req if an error occurred

File: main.go (hunk 1 above)
Comment (pb82) [reply main.go:112]: Yes, exactly

Comment (machi1990) [reply main.go:112]: It is reasonable. In an ideal scenario, we would have wished to not have to rewrite the req body again but we do not have a way out.

**Review by machi1990 — COMMENTED** [review]
the change lgtm

I'll let @valeriiashapoval have a look as well and approve.

**Comment by valeriiashapoval** [comment]
thanks @pb82 @machi1990 !

**Review by valeriiashapoval — APPROVED** [review]

**Pull request merged**